torchtune/models/clip/_position_embeddings.py

```diff
@@ -0,0 +1,167 @@
+        scale = embed_dim**-0.5
+        self.positional_embedding = nn.Parameter(
+            scale
+            * torch.randn((patch_grid_size**2 + 1, embed_dim))  # +1 for CLS token
+        )
+
+    def forward(self, x: torch.Tensor, *args) -> torch.Tensor:
+        """
+        args:
+            x (torch.Tensor): Tensor with shape (*, n_tokens, embed_dim)
+        returns:
+            torch.Tensor: The input tensor with added positional embeddings.
+        """
+        return x + self.positional_embedding
+
+
+class TiledTokenPositionalEmbedding(nn.Module):
+    """
+    Token positional embedding for tiled images. There are two positional embeddings in this module:
+        - local_token_positional_embedding: same for every tile, different for every token. Equivalent to
+            ``torchtune.models.clip._position_embeddings.TokenPositionalEmbedding``, but gated.
+        - global_token_positional_embedding: different for every tile, different for every token.
+
+    Notice that tile is different from patch (token). An image is divided into tiles during pre-processing,
+    and patches are the outcome of the convolution in the ViT applied to each tile.
+
+    Args:
+        max_num_tiles (int): The maximum number of tiles an image can be divided into.
+        embed_dim (int): The dimensionality of each token embedding.
+        patch_grid_size (int): The side of a squared grid that represents how many
+             patches are in one tile, i.e. tile_size // patch_size.
+    """
+
+    def __init__(
+        self, max_num_tiles: int, embed_dim: int, patch_grid_size: int
+    ) -> None:
+        super().__init__()
+        self.n_tokens_per_tile = patch_grid_size**2 + 1  # +1 for cls token
+        scale = embed_dim**-0.5
+
+        # different for every token, same for every tile
+        self.local_token_positional_embedding = nn.Parameter(
+            scale * torch.randn(patch_grid_size**2 + 1, embed_dim)
+        )
+
+        # different for every token, different for every tile
+        self.global_token_positional_embedding = nn.Parameter(
+            scale
+            * torch.randn(
+                max_num_tiles,
+                max_num_tiles,
+                self.n_tokens_per_tile,
+                embed_dim,
+            )
+        )
+
+        self.gate = nn.Parameter(torch.zeros(1))
+
+    def forward(self, x: torch.Tensor, aspect_ratio: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x (torch.Tensor): Tensor with shape (bsz, n_tiles, n_tokens, embed_dim)
+            aspect_ratio (torch.Tensor): Tensor with shape (bsz, 2), representing the aspect ratio of the image
+                before tile-cropping, e.g. (2,1).
+        Returns:
+            torch.Tensor: The input tensor with added positional embeddings.
+        """
+        # apply local position embedding (same for every tile)
+        bsz, n_tiles, n_tokens, embed_dim = x.shape
+
+        x = x.view(bsz * n_tiles, n_tokens, embed_dim)
```

Comment:
Sorry for being dense, but IIUC this view operation is needed to make the sum broadcastable - is that right?

Reply:
I dont think we actually need it! Will remove it. Good catch